File: ios/Plugin/SegmentAnalyticsPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(SegmentAnalyticsPlugin)
public class SegmentAnalyticsPlugin: CAPPlugin {
    private var initialized = false
    private let implementation = SegmentAnalytics()

    @objc func initialize(_ call: CAPPluginCall) {
        if initialized == true {
            call.reject("Segment is already initialized")
            return
        }
        guard let writeKey: String = call.getString("writeKey") else {
            call.reject("Write key is required to initialize plugin")
            return
        }
        let trackLifecycleEvents: Bool = call.getBool("trackLifecycleEvents") ?? false;
        let recordScreenViews: Bool = call.getBool("recordScreenViews") ?? false;
        implementation.initialize(writeKey: writeKey, trackLifecycleEvents: trackLifecycleEvents, recordScreenViews: recordScreenViews)
        initialized = true
        call.resolve()
    }

    @objc func identify(_ call: CAPPluginCall) {
        if initialized != true {
            call.reject("Segment is not initialized")
            return
        }
        guard let userId: String = call.getString("userId") else {
            call.reject("User ID is required for 'identify' but not supplied")
            return
        }

        let traits: Dictionary = call.getObject("traits") ?? [:]
        let options: Dictionary = call.getObject("options") ?? [:]
        implementation.identify(userId: userId, traits: traits, options: options)
        call.resolve()
    }

    @objc func track(_ call: CAPPluginCall) {
        if initialized != true {
            call.reject("Segment is not initialized")
            return
        }
        guard let eventName: String = call.getString("eventName") else {
            call.reject("Event name is not supplied")
            return
        }

        let properties: Dictionary = call.getObject("properties") ?? [:]
        let options: Dictionary = call.getObject("options") ?? [:]
        implementation.track(eventName: eventName, properties: properties, options: options)
        call.resolve()
    }

    @objc func screen(_ call: CAPPluginCall) {
        if initialized != true {
            call.reject("Segment is not initialized")
            return
        }

        guard let screenName: String = call.getString("screenName") else {
            call.reject("Screen name is not supplied")
            return
        }

        let category: String? = call.getString("category") ?? nil
        let properties: Dictionary = call.getObject("properties") ?? [:]
        let options: Dictionary = call.getObject("options") ?? [:]

        implementation.screen(screenName: screenName, category: category, properties: properties, options: options)
        call.resolve()
    }

    @objc func page(_ call: CAPPluginCall) {
        if initialized != true {
            call.reject("Segment is not initialized")
            return
        }

        guard let pageName: String = call.getString("pageName") else {
            call.reject("Page name is not supplied")
            return
        }

        let category: String? = call.getString("category") ?? nil
        let properties: Dictionary = call.getObject("properties") ?? [:]
        let options: Dictionary = call.getObject("options") ?? [:]

        implementation.screen(screenName: pageName, category: category, properties: properties, options: options)
        call.resolve()
    }

    @objc func group(_ call: CAPPluginCall) {
        if initialized != true {
            call.reject("Segment is not initialized")
            return
        }

        guard let userId: String = call.getString("userId") else {
            call.reject("User ID was not supplied")
            return
        }

        guard let groupId: String = call.getString("groupId") else {
            call.reject("Group ID was not supplied")
            return
        }

        let traits: Dictionary = call.getObject("traits") ?? [:]
        let options: Dictionary = call.getObject("options") ?? [:]
        implementation.group(userId: userId, groupId: groupId, traits: traits, options: options)
        call.resolve()
    }

    @objc func alias(_ call: CAPPluginCall) {
        if initialized != true {
            call.reject("Segment is not initialized")
            return
        }
        guard let newId: String = call.getString("newId") else {
            call.reject("New ID was not supplied")
            return
        }

        let options: Dictionary = call.getObject("options") ?? [:]
        implementation.alias(newId: newId, options: options)
        call.resolve()
    }

    @objc func reset(_ call: CAPPluginCall) {
        if initialized != true {
            call.reject("Segment is not initialized")
            return
        }

        implementation.reset()
        call.resolve()
    }
}

File: android/src/main/java/com/vjfuenzalida/capacitor/segment/SegmentAnalyticsPlugin.java
package com.vjfuenzalida.capacitor.segment;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "SegmentAnalytics")
public class SegmentAnalyticsPlugin extends Plugin {

    private boolean initialized = false;
    private SegmentAnalytics implementation = new SegmentAnalytics();

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void initialize(PluginCall call) {
        if (initialized == true) {
            call.reject("Segment is already initialized");
            return;
        }
        String writeKey = call.getString("writeKey");
        if (writeKey == null) {
            call.reject("Write key is required to initialize plugin");
            return;
        }
        Boolean trackLifecycleEvents = call.getBoolean("trackLifecycleEvents", false);
        Boolean recordScreenViews = call.getBoolean("recordScreenViews", false);

        implementation.initialize(writeKey, trackLifecycleEvents, recordScreenViews);
        initialized = true;
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void identify(PluginCall call) {
        if (!initialized) {
            call.reject("Segment is not initialized");
            return;
        }
        String userId = call.getString("userId");
        if (userId == null) {
            call.reject("User ID is required for identify");
            return;
        }
        JSObject traits = call.getObject("traits");
        JSObject options = call.getObject("options");

        implementation.identify(userId, traits, options);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void track(PluginCall call) {
        if (!initialized) {
            call.reject("Segment is not initialized");
            return;
        }
        String eventName = call.getString("eventName");
        if (eventName == null) {
            call.reject("Event name is not supplied");
            return;
        }
        JSObject properties = call.getObject("properties");
        JSObject options = call.getObject("options");

        implementation.track(eventName, properties, options);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void screen(PluginCall call) {
        if (!initialized) {
            call.reject("Segment is not initialized");
            return;
        }
        String screenName = call.getString("screenName");
        if (screenName == null) {
            call.reject("Screen name was not supplied");
            return;
        }
        JSObject properties = call.getObject("properties");
        JSObject options = call.getObject("options");

        implementation.screen(screenName, properties, options);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void page(PluginCall call) {
        if (!initialized) {
            call.reject("Segment is not initialized");
            return;
        }
        String pageName = call.getString("pageName");
        if (pageName == null) {
            call.reject("Page name was not supplied");
            return;
        }
        JSObject properties = call.getObject("properties");
        JSObject options = call.getObject("options");

        implementation.screen(pageName, properties, options);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void group(PluginCall call) {
        if (!initialized) {
            call.reject("Segment is not initialized");
            return;
        }
        String userId = call.getString("userId");
        String groupId = call.getString("groupId");
        if (userId == null || groupId == null) {
            call.reject("User ID and Group ID are required");
            return;
        }
        JSObject traits = call.getObject("traits");
        JSObject options = call.getObject("options");

        implementation.group(userId, groupId, traits, options);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void alias(PluginCall call) {
        if (!initialized) {
            call.reject("Segment is not initialized");
            return;
        }
        String newId = call.getString("newId");
        if (newId == null) {
            call.reject("New User ID is required");
            return;
        }
        JSObject options = call.getObject("options");

        implementation.group(newId, options);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void reset(PluginCall call) {
        if (!initialized) {
            call.reject("Segment is not initialized");
            return;
        }

        implementation.reset();
        call.resolve();
    }
}
